src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -167,34 +207,21 @@ private static String extractQualifiedName(DetailAST classExtend) {
     }
 
     /**
-     * Register to outer super classes of given classAst that
-     * given classAst is extending them.
-     * @param classAst class which outer super classes will be
-     *                 informed about nesting subclass
-     */
-    private void registerNestedSubclassToOuterSuperClasses(DetailAST classAst) {
-        final String currentAstSuperClassName = getSuperClassName(classAst);
-        if (currentAstSuperClassName != null) {
-            for (ClassDesc classDesc : classes) {
-                final String classDescQualifiedName = classDesc.getQualifiedName();
-                if (doesNameInExtendMatchSuperClassName(classDescQualifiedName,
-                        currentAstSuperClassName)) {
-                    classDesc.registerNestedSubclass();
-                }
-            }
-        }
-    }
-
-    /**
      * Get qualified class name from given class Ast.
      * @param classAst class to get qualified class name
      * @return qualified class name of a class
      */
     private String getQualifiedClassName(DetailAST classAst) {
         final String className = classAst.findFirstToken(TokenTypes.IDENT).getText();
         String outerClassQualifiedName = null;
-        if (!classes.isEmpty()) {
-            outerClassQualifiedName = classes.peek().getQualifiedName();
+        if (currentDepthLevel > 1) {
+            final Iterator<ClassDesc> iterator = classes.iterator();
+            while (outerClassQualifiedName == null) {
```

Comment:
@rnveach It depends on that we never cover the loop in your implementation because we will  never cover both loop's condition